File: for-ios-p2p/plugins/socket/ios/Sources/SocketPlugin/TcpSockets.swift
import Foundation
import Network

/**
 * TCP for the WebView.
 *
 * A WebView has fetch and WebSockets and nothing that speaks a raw stream, and
 * the Reaper transport is a raw stream: length-prefixed encrypted frames over
 * a socket that stays open. So the sockets live out here and the JavaScript
 * drives them by id.
 *
 * ## Why Network.framework
 *
 * Rather than BSD sockets, which would also work. Two reasons that matter on a
 * phone rather than a desktop:
 *
 *   - **It survives the network changing.** Walking out of the house moves the
 *     device from Wi-Fi to cellular, and `NWConnection` reports that as a state
 *     change instead of a socket that silently stops delivering.
 *   - **It is allowed to run in the background.** Raw sockets on a suspended
 *     app are closed by the system; these are managed and report their state
 *     honestly when the app comes back.
 *
 * ## Talking to Tor
 *
 * Every Reaper connection is to an onion address, which no TCP stack can
 * resolve. The address is handed to a SOCKS5 proxy instead — the Tor client
 * embedded in the app — and the handshake for that is implemented here rather
 * than in JavaScript, because it has to happen before the first byte of the
 * real protocol and the layer above should not have to know it happened.
 *
 * See `SOCKS5` below. It is the minimum: no authentication, domain-name
 * addressing, one reply.
 */

/** Bytes flowing one way, and the ids the WebView knows a socket by. */
final class TcpSockets {
  /** How a socket reports what happened to it. */
  typealias Events = (
    _ id: String,
    _ event: String,
    _ payload: [String: Any]
  ) -> Void

  private let queue = DispatchQueue(label: "chat.reaper.sockets")
  private var connections: [String: NWConnection] = [:]

  /**
   * Every listener, by the port it ended up bound to.
   *
   * There used to be one, and `listen` began by cancelling it. That is correct
   * for a device with one service and this device has two: the account address
   * forwards to the chat transport, and the sync address — the one a QR code
   * points at — forwards to the pairing service. They are separate onion
   * services with separate keys precisely so that a device can be reached for
   * linking whether or not it is answering for the account.
   *
   * `boot.ts` opens the pairing listener first and the transport second, so
   * the second call cancelled the first. Tor went on forwarding the sync
   * address to a port nothing was bound to any more, and every attempt to link
   * *to* this phone reached a closed port. From the other device that is
   * indistinguishable from the phone being switched off, which is why it read
   * as "linking only works in one direction".
   *
   * Keyed by bound port rather than by an id from the caller, because the port
   * is the thing Tor is configured with and therefore the thing an inbound
   * connection can actually be attributed to.
   */
  private var listeners: [UInt16: NWListener] = [:]

  /**
   * The order they were bound in, which is what makes a reload recoverable.
   *
   * The WebView can reload — linking a device does exactly that, because the
   * page has to come back up as somebody else — and it takes every listener
   * the JavaScript knew about with it. Tor does not reload: it goes on
   * forwarding both onion services to the ports it was told about at launch,
   * and `TorService.start` returns early when it is already running, so it
   * cannot be told new ones.
   *
   * So the ports that matter are the ones already bound here, and the fresh
   * JavaScript has to find them again rather than ask for two new ones nothing
   * forwards to. Order is how it tells them apart — `boot.ts` opens the
   * pairing listener first and the transport second, on every run — so this
   * preserves it. A dictionary alone does not.
   */
  private var order: [UInt16] = []

  private let emit: Events

  init(emit: @escaping Events) {
    self.emit = emit
  }

  // ---- outbound ------------------------------------------------------------

  /**
   * Open a connection through the local SOCKS5 proxy.
   *
   * `host` is an onion address. It is never resolved here — that is the entire
   * point of routing through Tor, and a DNS lookup for it would both fail and
   * tell the local resolver who is being contacted.
   */
  func connect(
    id: String,
    host: String,
    port: UInt16,
    proxyPort: UInt16
  ) {
    let endpoint = NWEndpoint.hostPort(
      host: .init("127.0.0.1"),
      port: .init(rawValue: proxyPort) ?? 9050
    )

    let options = NWProtocolTCP.Options()
    // The transport sends small frames that matter immediately — a keystroke
    // indicator, a call answer — and Nagle would hold them back to combine
    // them with whatever comes next.
    options.noDelay = true
    // Long enough to survive a quiet connection, short enough to notice a peer
    // that has gone away.
    options.enableKeepalive = true
    options.keepaliveIdle = 30

    let connection = NWConnection(
      to: endpoint,
      using: NWParameters(tls: nil, tcp: options)
    )

    connections[id] = connection

    connection.stateUpdateHandler = { [weak self] state in
      guard let self else { return }

      switch state {
      case .ready:
        // Connected to the proxy, not to the peer. The handshake is what turns
        // one into the other, and only then is the socket usable.
        self.socks5(id: id, connection: connection, host: host, port: port)

      case .failed(let error):
        self.emit(id, "error", ["message": error.localizedDescription])
        self.close(id: id)

      case .cancelled:
        self.emit(id, "close", [:])

      default:
        break
      }
    }

    connection.start(queue: queue)
  }

  // ---- inbound -------------------------------------------------------------

  /**
   * Accept connections from the Tor client.
   *
   * Bound to loopback only. An onion service is reached by Tor forwarding to a
   * local port, so nothing outside this device should ever be able to open
   * this — and binding it to every interface would put an unauthenticated
   * Reaper transport on the local Wi-Fi.
   */
  /**
   * Bind a loopback port and report which one was granted.
   *
   * Asynchronous, and it has to be. `NWListener.port` is nil until the
   * listener reaches `.ready`, and `start` returns immediately — so reading it
   * on the next line yields nothing and the caller falls back to the port it
   * asked for. Asking for zero means "choose one for me", so the caller was
   * told the port was zero, Tor was handed a forwarding target of zero, and
   * refused to start with "localPort is required".
   *
   * The symptom was "nothing can be sent or received", four layers away from a
   * value read one line too early.
   */
  func listen(port: UInt16, then: @escaping (Result<UInt16, Error>) -> Void) {
    // A specific port asked for twice is the same listener asked for twice —
    // `PairService.open()` and `transport.listen()` are both idempotent on the
    // JavaScript side and may be reached more than once. Only an exact repeat
    // is treated that way; asking for zero always binds something new.
    if port != 0, listeners[port] != nil {
      then(.success(port))
      return
    }

    let options = NWProtocolTCP.Options()
    options.noDelay = true

    let parameters = NWParameters(tls: nil, tcp: options)
    parameters.allowLocalEndpointReuse = true

    // Only pinned when a specific port was asked for. Requiring an endpoint of
    // port zero asks the system to bind port zero literally, rather than to
    // pick a free one.
    if port != 0 {
      parameters.requiredLocalEndpoint = NWEndpoint.hostPort(
        host: .init("127.0.0.1"),
        port: .init(rawValue: port) ?? .any
      )
    }

    let listener: NWListener
    do {
      listener = try NWListener(using: parameters)
    } catch {
      then(.failure(error))
      return
    }

    // Answered once, whichever way it goes.
    var answered = false
    let settle: (Result<UInt16, Error>) -> Void = { outcome in
      if answered { return }
      answered = true
      then(outcome)
    }

    listener.stateUpdateHandler = { [weak self] state in
      switch state {
      case .ready:
        guard let bound = listener.port?.rawValue else {
          settle(.failure(NSError(
            domain: "chat.reaper.sockets", code: 1,
            userInfo: [NSLocalizedDescriptionKey: "listener is ready but reported no port"]
          )))
          return
        }

        // Recorded under the port it actually got, which for a request of zero
        // is not known until now. `stopListening` and the accept handler both
        // work from this map, so registering late is the only correct moment:
        // a listener filed under the port that was *asked* for would be filed
        // under zero.
        self?.listeners[bound] = listener
        if self?.order.contains(bound) == false { self?.order.append(bound) }

        settle(.success(bound))

      case .failed(let error):
        settle(.failure(error))

      case .cancelled:
        settle(.failure(NSError(
          domain: "chat.reaper.sockets", code: 2,
          userInfo: [NSLocalizedDescriptionKey: "listener was cancelled before it bound"]
        )))

      case .waiting(let reason):
        // Not an error, and that is the danger: Network.framework means "the
        // port is busy, I will keep trying", and on a loopback port that is
        // never freed it waits for ever. Nothing settles, and every `await`
        // above this one stops permanently — which presents as an app frozen
        // on its startup screen with no error anywhere.
        //
        // Treated as a failure, because for this caller it is one: the port
        // was asked for and cannot be had.
        settle(.failure(NSError(
          domain: "chat.reaper.sockets", code: 3,
          userInfo: [NSLocalizedDescriptionKey:
            "could not bind the port: \(reason.localizedDescription)"]
        )))

      default:
        break
      }
    }

    listener.newConnectionHandler = { [weak self] connection in
      guard let self else { return }

      let id = "in-\(UUID().uuidString)"
      self.connections[id] = connection

      // Which local port this arrived on, so the JavaScript side can hand it
      // to the right server. Two onion services forward to two ports here and
      // they speak different protocols — chat frames and pairing frames — so a
      // connection given to the wrong one is not merely misrouted. It is read
      // as a corrupt length prefix and the socket is destroyed, which is the
      // "that device closed the connection" this app has produced before from
      // a different cause.
      //
      // Read off the listener rather than the connection: `localEndpoint` on
      // an inbound NWConnection is not populated on every path, and the
      // listener's port is the value Tor was actually configured with.
      let arrivedOn = Int(listener.port?.rawValue ?? 0)

      connection.stateUpdateHandler = { state in
        switch state {
        case .ready:
          self.emit(id, "accept", ["port": arrivedOn])
          self.read(id: id, connection: connection)
        case .failed(let error):
          self.emit(id, "error", ["message": error.localizedDescription])
          self.close(id: id)
        case .cancelled:
          self.emit(id, "close", [:])
        default:
          break
        }
      }

      connection.start(queue: self.queue)
    }

    listener.start(queue: queue)
  }

  /** Stop one listener, or every one of them. */
  func stopListening(port: UInt16 = 0) {
    if port == 0 {
      for listener in listeners.values { listener.cancel() }
      listeners.removeAll()
      order.removeAll()
      return
    }

    listeners.removeValue(forKey: port)?.cancel()
    order.removeAll { $0 == port }
  }

  /**
   * The ports currently bound, oldest first.
   *
   * Read by the JavaScript after a reload so it can re-attach to the listeners
   * Tor is still forwarding to, instead of binding two new ports that nothing
   * points at. Without it, linking a device left the phone reachable only
   * until the page reloaded — which linking always does — and it stayed
   * unreachable until the app was force-quit.
   */
  func listeningPorts() -> [UInt16] {
    order.filter { listeners[$0] != nil }
  }

  // ---- moving bytes --------------------------------------------------------

  func send(id: String, data: Data) {
    guard let connection = connections[id] else { return }

    connection.send(
      content: data,
      completion: .contentProcessed { [weak self] error in
        guard let error else { return }
        self?.emit(id, "error", ["message": error.localizedDescription])
        self?.close(id: id)
      }
    )
  }

  /**
   * Read forever, handing each chunk up as it arrives.
   *
   * No framing here. The transport already length-prefixes its own frames and
   * has to cope with a chunk that is half a frame or three of them, because
   * that is what TCP does — reimplementing that boundary logic in Swift would
   * be a second copy of it to keep in step.
   */
  private func read(id: String, connection: NWConnection) {
    connection.receive(
      minimumIncompleteLength: 1,
      maximumLength: 64 * 1024
    ) { [weak self] data, _, complete, error in
      guard let self else { return }

      if let data, !data.isEmpty {
        self.emit(id, "data", ["data": data.base64EncodedString()])
      }

      if let error {
        self.emit(id, "error", ["message": error.localizedDescription])
        self.close(id: id)
        return
      }

      if complete {
        self.close(id: id)
        return
      }

      self.read(id: id, connection: connection)
    }
  }

  func close(id: String) {
    guard let connection = connections.removeValue(forKey: id) else { return }
    connection.cancel()
  }

  func closeAll() {
    for id in connections.keys { close(id: id) }
    stopListening()
  }

  // ---- SOCKS5 --------------------------------------------------------------
  //
  // RFC 1928, the subset Tor implements and this needs:
  //
  //   greeting   05 01 00                  version, one method, "no auth"
  //   choice     05 00                      version, method accepted
  //   request    05 01 00 03 len host port  connect, by domain name
  //   reply      05 00 00 ...               version, success, bound address
  //
  // The domain-name address type is the important one: it hands the onion
  // address to Tor as a name and lets Tor do the resolving, which is what
  // keeps the destination off this device's DNS.

  private func socks5(
    id: String,
    connection: NWConnection,
    host: String,
    port: UInt16
  ) {
    let greeting = Data([0x05, 0x01, 0x00])

    connection.send(content: greeting, completion: .contentProcessed { [weak self] error in
      guard let self else { return }

      if let error {
        self.emit(id, "error", ["message": "proxy: \(error.localizedDescription)"])
        self.close(id: id)
        return
      }

      self.expect(id: id, connection: connection, bytes: 2) { choice in
        guard choice.count == 2, choice[0] == 0x05, choice[1] == 0x00 else {
          self.emit(id, "error", ["message": "proxy refused the connection"])
          self.close(id: id)
          return
        }

        self.socks5Connect(id: id, connection: connection, host: host, port: port)
      }
    })
  }

  private func socks5Connect(
    id: String,
    connection: NWConnection,
    host: String,
    port: UInt16
  ) {
    guard let name = host.data(using: .utf8), name.count <= 255 else {
      emit(id, "error", ["message": "address too long"])
      close(id: id)
      return
    }

    var request = Data([0x05, 0x01, 0x00, 0x03, UInt8(name.count)])
    request.append(name)
    request.append(UInt8(port >> 8))
    request.append(UInt8(port & 0xff))

    connection.send(content: request, completion: .contentProcessed { [weak self] error in
      guard let self else { return }

      if let error {
        self.emit(id, "error", ["message": "proxy: \(error.localizedDescription)"])
        self.close(id: id)
        return
      }

      // The reply header is four bytes plus a bound address whose length
      // depends on its type, so it is read in two parts rather than guessed.
      self.expect(id: id, connection: connection, bytes: 4) { head in
        guard head.count == 4, head[0] == 0x05 else {
          self.emit(id, "error", ["message": "bad proxy reply"])
          self.close(id: id)
          return
        }

        guard head[1] == 0x00 else {
          self.emit(id, "error", ["message": Self.socksError(head[1])])
          self.close(id: id)
          return
        }

        let trailing: Int
        switch head[3] {
        case 0x01: trailing = 4 + 2            // IPv4 and port
        case 0x04: trailing = 16 + 2           // IPv6 and port
        case 0x03: trailing = -1               // length-prefixed name
        default:
          self.emit(id, "error", ["message": "unknown address type"])
          self.close(id: id)
          return
        }

        let finish = {
          self.emit(id, "connect", [:])
          self.read(id: id, connection: connection)
        }

        if trailing >= 0 {
          self.expect(id: id, connection: connection, bytes: trailing) { _ in finish() }
        } else {
          self.expect(id: id, connection: connection, bytes: 1) { length in
            guard let count = length.first else {
              self.close(id: id)
              return
            }
            self.expect(id: id, connection: connection, bytes: Int(count) + 2) { _ in
              finish()
            }
          }
        }
      }
    })
  }

  /**
   * Read exactly this many bytes.
   *
   * `receive(minimumIncompleteLength:maximumLength:)` with both set to the same
   * value will not return short, which is what the handshake needs — every
   * field in it is fixed-length and reading one byte too few desynchronises
   * everything after.
   */
  private func expect(
    id: String,
    connection: NWConnection,
    bytes: Int,
    then: @escaping (Data) -> Void
  ) {
    connection.receive(
      minimumIncompleteLength: bytes,
      maximumLength: bytes
    ) { [weak self] data, _, _, error in
      guard let self else { return }

      if let error {
        self.emit(id, "error", ["message": "proxy: \(error.localizedDescription)"])
        self.close(id: id)
        return
      }

      guard let data, data.count == bytes else {
        self.emit(id, "error", ["message": "proxy closed during handshake"])
        self.close(id: id)
        return
      }

      then(data)
    }
  }

  private static func socksError(_ code: UInt8) -> String {
    switch code {
    case 0x01: return "proxy failed"
    case 0x02: return "not allowed"
    case 0x03: return "network unreachable"
    case 0x04: return "host unreachable — the onion service may be offline"
    case 0x05: return "connection refused"
    case 0x06: return "timed out"
    default: return "proxy error \(code)"
    }
  }
}

File: for-ios-p2p/plugins/socket/android/src/main/java/chat/reaper/socket/TcpSockets.kt
package chat.reaper.socket

import android.util.Base64
import com.getcapacitor.JSObject
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

/**
 * TCP for the WebView, on Android.
 *
 * The counterpart to iOS's `TcpSockets.swift`, which uses `Network.framework`
 * for two reasons that do not apply here: surviving a Wi-Fi/cellular handoff,
 * and staying alive in the background. A plain `java.net.Socket` held by a
 * foreground service (see the keepalive plugin) already survives
 * backgrounding — Android's suspension model only kills a process, not a
 * socket a running service holds — so there is no framework-managed
 * connection type to reach for; blocking sockets on background coroutines are
 * enough. Wi-Fi/cellular handoff is not specially handled either: a dead
 * connection surfaces as a read/write `IOException` the same as any other
 * failure, and the JS side already retries through the outbox mechanism.
 *
 * Sockets are addressed by id rather than handed across as objects, because
 * only JSON crosses the bridge. Payloads are base64 for the same reason.
 */
class TcpSockets(private val emit: (id: String, event: String, payload: JSObject) -> Unit) {

    /** A socket plus the lock that keeps two overlapping `send()`s from interleaving. */
    private class Conn(val socket: Socket) {
        val writeLock = Any()
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val connections = ConcurrentHashMap<String, Conn>()

    /**
     * Every listener, by the port it ended up bound to.
     *
     * This device runs two onion services — the account address, forwarding
     * to the chat transport, and the sync address, forwarding to the pairing
     * service — and both have to stay bound at once. Keyed by port, not by a
     * caller-supplied id, because the port is what Tor is configured with and
     * therefore the only thing an inbound connection can be attributed to.
     */
    private val listeners = ConcurrentHashMap<Int, ServerSocket>()

    /**
     * The order listeners were bound in, which is what makes a WebView reload
     * recoverable. Tor does not reload — it goes on forwarding both onion
     * services to the ports it was told about at launch — so the fresh page
     * has to re-attach to the listeners already here (`listeningPorts()`)
     * rather than bind two new ones nothing forwards to.
     */
    private val order = CopyOnWriteArrayList<Int>()

    // ---- outbound --------------------------------------------------------

    /**
     * Open a connection through the local SOCKS5 proxy.
     *
     * `host` is an onion address. It is never resolved here — that is the
     * entire point of routing through Tor, and a DNS lookup for it would both
     * fail and tell the local resolver who is being contacted.
     */
    fun connect(id: String, host: String, port: Int, proxyPort: Int) {
        scope.launch {
            val socket = Socket()
            val conn = Conn(socket)
            connections[id] = conn

            try {
                // Small frames matter immediately here — a keystroke
                // indicator, a call answer — so Nagle's algorithm, which
                // would hold them back to combine with whatever comes next,
                // is switched off.
                socket.tcpNoDelay = true
                socket.keepAlive = true
                // Loopback to the embedded proxy: near-instant when it
                // succeeds, and bounded so a proxy that never came up fails
                // fast rather than hanging a coroutine forever.
                socket.connect(InetSocketAddress("127.0.0.1", proxyPort), 10_000)
                socks5Connect(conn, host, port)
            } catch (e: Exception) {
                if (connections.containsKey(id)) {
                    emit(id, "error", JSObject().put("message", describeError(e)))
                }
                close(id)
                return@launch
            }

            // Connected to the proxy *and* through the handshake to the
            // peer. Only now is the socket usable.
            emit(id, "connect", JSObject())
            readLoop(id, conn)
        }
    }

    // ---- inbound -----------------------------------------------------------

    /**
     * Bind a loopback port and report which one was granted.
     *
     * Bound to loopback only. An onion service is reached by Tor forwarding
     * to a local port, so nothing outside this device should ever be able to
     * open this — binding every interface would put an unauthenticated
     * Reaper transport on the local Wi-Fi.
     */
    fun listen(port: Int, then: (Result<Int>) -> Unit) {
        // A specific port asked for twice is the same listener asked for
        // twice — callers on the JS side are idempotent and may ask more
        // than once. Only an exact repeat is treated that way; asking for
        // zero always binds something new.
        if (port != 0 && listeners.containsKey(port)) {
            then(Result.success(port))
            return
        }

        scope.launch {
            val server: ServerSocket
            try {
                server = ServerSocket()
                server.reuseAddress = true
                server.bind(InetSocketAddress("127.0.0.1", port))
            } catch (e: IOException) {
                then(Result.failure(e))
                return@launch
            }

            val bound = server.localPort
            // Recorded under the port it actually got, which for a request
            // of zero is not known until the bind completes.
            listeners[bound] = server
            if (!order.contains(bound)) order.add(bound)

            then(Result.success(bound))
            acceptLoop(bound, server)
        }
    }

    private fun acceptLoop(boundPort: Int, server: ServerSocket) {
        while (!server.isClosed) {
            val socket = try {
                server.accept()
            } catch (e: IOException) {
                // Normal on close(): accept() unblocks with an exception the
                // instant the listening socket is closed elsewhere.
                break
            }

            val id = "in-" + UUID.randomUUID().toString()
            val conn = Conn(socket)
            connections[id] = conn
            try {
                socket.tcpNoDelay = true
            } catch (e: IOException) {
                // Best-effort; the connection is still usable without it.
            }

            // Which local port this arrived on, so the JS side can hand it to
            // the right server. Two onion services forward to two ports and
            // they speak different protocols — a connection given to the
            // wrong one reads as a corrupt length prefix, not a routing bug.
            emit(id, "accept", JSObject().put("port", boundPort))
            scope.launch { readLoop(id, conn) }
        }
    }

    /** Stop one listener, or every one of them. */
    fun stopListening(port: Int) {
        if (port == 0) {
            listeners.values.forEach { safeClose(it) }
            listeners.clear()
            order.clear()
            return
        }

        listeners.remove(port)?.let { safeClose(it) }
        order.remove(port)
    }

    /**
     * The ports currently bound, oldest first.
     *
     * Read by the JS side after a reload so it can re-attach to the listeners
     * Tor is still forwarding to, instead of binding two new ports nothing
     * points at.
     */
    fun listeningPorts(): List<Int> = order.filter { listeners.containsKey(it) }

    // ---- moving bytes ------------------------------------------------------

    fun send(id: String, data: ByteArray) {
        val conn = connections[id] ?: return
        scope.launch {
            try {
                synchronized(conn.writeLock) {
                    conn.socket.getOutputStream().write(data)
                    conn.socket.getOutputStream().flush()
                }
            } catch (e: IOException) {
                emit(id, "error", JSObject().put("message", describeError(e)))
                close(id)
            }
        }
    }

    /**
     * Read forever, handing each chunk up as it arrives.
     *
     * No framing here. The transport already length-prefixes its own frames
     * and copes with a chunk that is half a frame or three of them, because
     * that is what TCP does — reimplementing that boundary logic here would
     * be a second copy of it to keep in step.
     */
    private fun readLoop(id: String, conn: Conn) {
        val input = conn.socket.getInputStream()
        val buf = ByteArray(64 * 1024)

        try {
            while (true) {
                val n = input.read(buf)
                if (n == -1) {
                    close(id)
                    return
                }
                if (n > 0) {
                    val chunk = Base64.encodeToString(buf, 0, n, Base64.NO_WRAP)
                    emit(id, "data", JSObject().put("data", chunk))
                }
            }
        } catch (e: IOException) {
            if (connections.containsKey(id)) {
                emit(id, "error", JSObject().put("message", describeError(e)))
            }
            close(id)
        }
    }

    /** Idempotent: a second call on an id already gone is a silent no-op. */
    fun close(id: String) {
        val conn = connections.remove(id) ?: return
        safeClose(conn.socket)
        emit(id, "close", JSObject())
    }

    fun closeAll() {
        connections.keys.toList().forEach { close(it) }
        stopListening(0)
    }

    // ---- SOCKS5 --------------------------------------------------------------
    //
    // RFC 1928, the subset Tor implements and this needs:
    //
    //   greeting   05 01 00                  version, one method, "no auth"
    //   choice     05 00                      version, method accepted
    //   request    05 01 00 03 len host port  connect, by domain name
    //   reply      05 00 00 ...               version, success, bound address
    //
    // The domain-name address type is the important one: it hands the onion
    // address to Tor as a name and lets Tor do the resolving, which is what
    // keeps the destination off this device's DNS.

    private fun socks5Connect(conn: Conn, host: String, port: Int) {
        val socket = conn.socket
        val input = socket.getInputStream()

        writeLocked(conn, byteArrayOf(0x05, 0x01, 0x00))
        val choice = readExact(input, 2)
        if (choice[0] != 0x05.toByte() || choice[1] != 0x00.toByte()) {
            throw IOException("proxy refused the connection")
        }

        val name = host.toByteArray(Charsets.UTF_8)
        if (name.size > 255) throw IOException("address too long")

        val request = ByteArrayOutputStream()
        request.write(byteArrayOf(0x05, 0x01, 0x00, 0x03, name.size.toByte()))
        request.write(name)
        request.write((port shr 8) and 0xff)
        request.write(port and 0xff)
        writeLocked(conn, request.toByteArray())

        // The reply header is four bytes plus a bound address whose length
        // depends on its type, so it is read in two parts rather than guessed.
        val head = readExact(input, 4)
        if (head[0] != 0x05.toByte()) throw IOException("bad proxy reply")
        if (head[1] != 0x00.toByte()) throw IOException(socksError(head[1]))

        val trailing = when (head[3].toInt() and 0xff) {
            0x01 -> 4 + 2   // IPv4 and port
            0x04 -> 16 + 2  // IPv6 and port
            0x03 -> -1      // length-prefixed name
            else -> throw IOException("unknown address type")
        }

        if (trailing >= 0) {
            readExact(input, trailing)
        } else {
            val length = readExact(input, 1)[0].toInt() and 0xff
            readExact(input, length + 2)
        }
    }

    private fun writeLocked(conn: Conn, bytes: ByteArray) {
        synchronized(conn.writeLock) {
            conn.socket.getOutputStream().write(bytes)
            conn.socket.getOutputStream().flush()
        }
    }

    /**
     * Read exactly this many bytes, or throw.
     *
     * Every field in the handshake is fixed-length, and reading one byte too
     * few would desynchronise everything after it — so this loops rather than
     * trusting a single `read()` to fill the buffer, which it is not obliged
     * to do.
     */
    private fun readExact(input: InputStream, n: Int): ByteArray {
        if (n == 0) return ByteArray(0)
        val buf = ByteArray(n)
        var off = 0
        while (off < n) {
            val r = input.read(buf, off, n - off)
            if (r == -1) throw IOException("proxy closed during handshake")
            off += r
        }
        return buf
    }

    private fun socksError(code: Byte): String = when (code.toInt() and 0xff) {
        0x01 -> "proxy failed"
        0x02 -> "not allowed"
        0x03 -> "network unreachable"
        0x04 -> "host unreachable — the onion service may be offline"
        0x05 -> "connection refused"
        0x06 -> "timed out"
        else -> "proxy error ${code.toInt() and 0xff}"
    }

    private fun describeError(e: Exception): String = e.message ?: e.toString()

    private fun safeClose(socket: Socket) {
        try { socket.close() } catch (e: IOException) { /* already closed */ }
    }

    private fun safeClose(server: ServerSocket) {
        try { server.close() } catch (e: IOException) { /* already closed */ }
    }
}
